cpp/data_structures/dynamic_tree/DynamicTree.hpp

```diff
@@ -0,0 +1,441 @@
+﻿#include <iostream>
+#include <algorithm>
+#include <memory>
+#include <string>
+#include <vector>
+
+#pragma once
+
+#ifndef DYNAMIC_TREE_H
+#define DYNAMIC_TREE_H
+
+namespace cpstl {
+template <class T>
+class DynamicTree;
+
+// ===================
+// === Nodes class ===
+// ===================
+template <typename T>
+class DTNode {
+ private:
+  // parent of this node
+  std::shared_ptr<DTNode<T>> parent;
+
+  // node name => each node must have a name for dealing with it => "search ,
+  // sort , ..."
+  std::string name;
+
+  // o(log n)
+  // search for children in temp node
+  int binary_search(std::string const& target_name) {
+    int min = 0;
+    int max = this->children.size() - 1;
+    int mid = (min + max) / 2;
+
+    while (min <= max) {
+      // in case target found
+      if (this->children[mid]->name == target_name) {
+        return mid;
+      }
+
+      // going left
+      if (this->children[mid]->name.compare(target_name) > 0) {
+        max = (mid - 1);
+      }
+
+      // going right
+      if (this->children[mid]->name.compare(target_name) < 0) {
+        min = (mid + 1);
+      }
+
+      // update mid
+      mid = (min + max) / 2;
+    }
+
+    // in case not found
+    return -1;
+  }
+
+  bool sub_full_search(std::string const& target_name) {
+    if (this->name == target_name)
+      return true;
+    else {
+      int index = this->binary_search(target_name);
+      if (index != -1) return true;
+
+      for (std::shared_ptr<DTNode<T>> node : this->children) {
+        if (node->sub_full_search(target_name)) return true;
+      }
+    }
+
+    return false;
+  }
+
+ public:
+  // children of this node
+  std::vector<std::shared_ptr<DTNode<T>>> children;
+
+  // node value
+  T value;
+
+  // default constructor
+  DTNode() {}
+
+  DTNode(std::string node_name, T node_value,
+         std::shared_ptr<DTNode<T>> parent_node = NULL)
+      : name{node_name}, value{node_value}, parent{parent_node} {}
+
+  std::string get_name() { return this->name; }
+
+  std::string get_parent_name() { return this->parent->get_name(); }
+
+  // get length of children in this node
+  std::size_t length() { return this->children.size(); }
+
+  // remove all children of "current node"
+  void remove_children() {
+    // nested recursive remove_children => remove all children of children
+    for (std::shared_ptr<DTNode<T>>& child : this->children) {
+      child->remove_children();
+    }
+
+    // last step make a new empty vector
+    this->children = std::vector<std::shared_ptr<DTNode<T>>>();
+  }
+
+  // o(1)
```

Comment:
```suggestion
  // O(1)
```

You are sure that the copy is O(1), you need to copy a sequence of bytes right? O(N)